### ISSUE TITLE
Defining the neighborhood of a datapoint

### ISSUE THREAD

**Issue by ajayaadhikari** [marcotcr/lime#185]
In the paper artificial data-points are sampled in the neighborhood of the instance being explained. This neighborhood is defined by an exponential kernel defined on some distance function D. I have some questions about it.
Why is an exponential kernel chosen?
How should D be chosen?

Because we do not know how the black-box model defines the distance. E.g. a categorical feature with vales b,c,d,e. Going from b to c might be small but going form b to d might be huge. Even with normalized numerical features the same difference of two numerical features might be very different in the eye of the black-box model. I think by choosing different values of D the explanations can be very different.
I would appreciate any feedback or thoughts.

**Comment by ajayaadhikari** [comment]
Some background for my question:
I am doing research in providing examples from the training set that are similar to the instance being explained in local model-agnostic setting. Similar in the sense that the black-box classifier would classify it as the same class for similar reasons. I am thinking to use LIME explanations as kernel to compute the distance.
But in the paper exponential kernel with a distance function D is used to define the neighborhood.
So do you guys think that using LIME explanations as kernel or the kernel from the paper directly would yield the same result?

**Comment by marcotcr** [comment]
> Why is an exponential kernel chosen?

I guess the choice was a bit arbitrary, but it is a kernel that has nice intuitive properties (e.g. weight decaying exponentially w.r.t. distance)
> How should D be chosen?

D has to be a distance function that is intuitive to the user, i.e. the user should be able to grasp if two examples are far or close given the intuitive representation. This may be challenging, which is part of the reason why we wrote [this paper](http://homes.cs.washington.edu/~marcotcr/aaai16.pdf).

> Because we do not know how the black-box model defines the distance. E.g. a categorical feature with vales b,c,d,e. Going from b to c might be small but going form b to d might be huge. Even with normalized numerical features the same difference of two numerical features might be very different in the eye of the black-box model. I think by choosing different values of D the explanations can be very different.

You are right that we don't know about the black box model, but I think this is for the best as we want the explanation to really be model agnostic. We do get an idea of how the black box model ranks features, by perturbing them and seeing how it reacts. However, the point you raise is an important one: going from b to c may not be the same as going from b to e from the user's perspective, and our implementation assumes it is the same.
Also you are correct that the explanation is with respect to the distance function D and the interpretable representation.

> I am doing research in providing examples from the training set that are similar to the instance being explained in local model-agnostic setting. Similar in the sense that the black-box classifier would classify it as the same class for similar reasons. I am thinking to use LIME explanations as kernel to compute the distance.
> But in the paper exponential kernel with a distance function D is used to define the neighborhood.
> So do you guys think that using LIME explanations as kernel or the kernel from the paper directly would yield the same result?
> 

I think what you mean is using the LIME explanations as a representation, rather than as a kernel.
If that is what you mean, I don't think that you'll get the same results if you just rank examples by the distance function (or by the exponential kernel on the distance function, which yields the same ranking in reverse). Let's take a simple example with two numerical features: height and hair length. Let's pretend that they are independent, and that I'm trying to predict if a person is heavier than 150 (i.e. only height is going to help). Assuming the model only cares about height, LIME explanations will put a lot of weight on the 'height' feature and no weight on 'hair length'. To compute the closest instances to the instance being explained using just euclidean distance, hair length will be taken into account and thus you will get different results.

**Comment by ajayaadhikari** [comment]
Thank you for your elaborate answer.
> I think what you mean is using the LIME explanations as a representation, rather than as a kernel.
If that is what you mean

Yes, I mean as representation
> Assuming the model only cares about height, LIME explanations will put a lot of weight on the 'height' feature and no weight on 'hair length'. Examples closest to the one being explained using LIME explanations as representations will be people with most similar weight. To compute the closest instances to the instance being explained using just euclidean distance, hair length will be taken into account and thus you will get different results.

You have a point that only using the distance function does not look into the black-box, and it will give equal weights to "height" and "hair length", while LIME explanations will put a lot of weight on the "height" feature which is according to the black-box. But in this example the model thinks globally that "hair length" does not matter so any neighborhood will yield this result regardless of the real neighborhood of the instance being explained. But let's say if "hair length" only matters in the neighborhood of instance being explained, then the distance function of defining the neighborhood is crucial. Using an arbitrary distance function might not yield the real neighborhood according to the black-box consequently the LIME explanation will not give importance to "hair length".
I think that a custom distance measure according to the black-box is needed to have a local explanation with high fidelity to the black-box. What are your thoughts on this?

**Comment by marcotcr** [comment]
> But in this example the model thinks globally that "hair length" does not matter so any neighborhood will yield this result regardless of the real neighborhood of the instance being explained.


I don't know what you mean by 'real neighborhood' here.

> But let's say if "hair length" only matters in the neighborhood of instance being explained, then the distance function of defining the neighborhood is crucial. Using an arbitrary distance function might not yield the real neighborhood according to the black-box consequently the LIME explanation will not give importance to "hair length".

The distance function obviously influences what explanations LIME will find, as it will weight instances according to it. But in this example, LIME will not give 'hair length' a lot of weight only if examples that show the effect of hair length are really far according to the distance function (e.g. if the distance function puts infinite distance between points with different hair length). However, I don't see this being a problem with standard distance functions (e.g. euclidean, cosine).

> I think that a custom distance measure according to the black-box is needed to have a local explanation with high fidelity to the black-box. What are your thoughts on this?

I don't understand why this would be, as LIME is already taking the black box into account by trying to approximate its predictions.

**Comment by ajayaadhikari** [comment]
Thank you for your answer.
Sorry for a late reply.

> I don't know what you mean by 'real neighborhood' here.

By real neighborhood, I mean the neighborhood defined by the distance function in the representation used by the black-box model. 

> You are right that we don't know about the black box model, but I think this is for the best as we want the explanation to really be model agnostic.
> ... Also you are correct that the explanation is with respect to the distance function D and the interpretable representation.

My point is that the distance function/representation used by LIME and the distance function/representation used by the underlying black-box model could be totally different as you said. In that case is it valid to say that the explanation is locally faithful to the underlying model?
In other words, the local fidelity of an explanation could vary according to whether the representation/distance used by LIME is similar to representation/distance used by the underlying model.
This raises a fundamental question whether any local explanation method can be both model agnostic and have a high local fidelity to the underlying model because we do not know the distance/representation of the underlying model.

**Comment by marcotcr** [comment]
Maybe we are operating under different definitions of 'high fidelity'. I would consider an explanation to be faithful under some perturbation distribution / distance function if it predicts the same as the underlying model, even if the model uses a different representation, or weights things differently.
Good question : )
